# <p style="padding:10px;background: linear-gradient(to right, #ab70c6, #4a006c);margin:0;color:#000000;font-family:newtimeroman;font-size:150%;text-align:center;border-radius: 15px 50px;overflow:hidden;font-weight:500">📹🍿 Movies Recommendation System 📊✨
</p>


<p align="center">
  <img src="Interstellar_photo.jpg" alt="Project Banner" width="1000"/>
</p>

<a id="1"></a>
<p style="padding:10px;
          background: linear-gradient(to right, #ab70c6, #4a006c);
          margin:0;
          color:#000000;
          font-family:newtimeroman;
          font-size:250%;
          text-align:center;
          border-radius: 15px 50px;
          overflow:hidden;
          font-weight:500">
Importing Libraries
</p>


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

<a id="1"></a>
<p style="padding:10px;
          background: linear-gradient(to right, #ab70c6, #4a006c);
          margin:0;
          color:#000000;
          font-family:newtimeroman;
          font-size:250%;
          text-align:center;
          border-radius: 15px 50px;
          overflow:hidden;
          font-weight:500">
Loading datasets
</p>


In [2]:
credits = pd.read_csv(r'tmdb_5000_credits.csv')
credits.head()

,movie_id,title,cast,crew
0,19995,Avatar,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,206647,Spectre,"[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."
3,49026,The Dark Knight Rises,"[{""cast_id"": 2, ""character"": ""Bruce Wayne / Ba...","[{""credit_id"": ""52fe4781c3a36847f81398c3"", ""de..."
4,49529,John Carter,"[{""cast_id"": 5, ""character"": ""John Carter"", ""c...","[{""credit_id"": ""52fe479ac3a36847f813eaa3"", ""de..."


In [3]:
movies = pd.read_csv('tmdb_5000_movies.csv')
movies.head(3)

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.sonypictures.com/movies/spectre/,206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...","[{""iso_3166_1"": ""GB"", ""name"": ""United Kingdom""...",2015-10-26,880674609,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,Spectre,6.3,4466


<a id="1"></a>
<p style="padding:10px;
          background: linear-gradient(to right, #ab70c6, #4a006c);
          margin:0;
          color:#000000;
          font-family:newtimeroman;
          font-size:250%;
          text-align:center;
          border-radius: 15px 50px;
          overflow:hidden;
          font-weight:500">
Data Preparation
</p>


In [4]:
movies.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count'],
      dtype='object')

In [5]:
credits.columns

Index(['movie_id', 'title', 'cast', 'crew'], dtype='object')

In [6]:
# rename movie_id column in credits dataset to can merge with movies dataset
# rename new data to credits_updated_df

credits_updated_df = credits.rename(columns={'movie_id': 'id'})
credits_updated_df

,id,title,cast,crew
0,19995,Avatar,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,206647,Spectre,"[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."
3,49026,The Dark Knight Rises,"[{""cast_id"": 2, ""character"": ""Bruce Wayne / Ba...","[{""credit_id"": ""52fe4781c3a36847f81398c3"", ""de..."
4,49529,John Carter,"[{""cast_id"": 5, ""character"": ""John Carter"", ""c...","[{""credit_id"": ""52fe479ac3a36847f813eaa3"", ""de..."
...,...,...,...,...
4798,9367,El Mariachi,"[{""cast_id"": 1, ""character"": ""El Mariachi"", ""c...","[{""credit_id"": ""52fe44eec3a36847f80b280b"", ""de..."
4799,72766,Newlyweds,"[{""cast_id"": 1, ""character"": ""Buzzy"", ""credit_...","[{""credit_id"": ""52fe487dc3a368484e0fb013"", ""de..."
4800,231617,"Signed, Sealed, Delivered","[{""cast_id"": 8, ""character"": ""Oliver O\u2019To...","[{""credit_id"": ""52fe4df3c3a36847f8275ecf"", ""de..."
4801,126186,Shanghai Calling,"[{""cast_id"": 3, ""character"": ""Sam"", ""credit_id...","[{""credit_id"": ""52fe4ad9c3a368484e16a36b"", ""de..."


In [7]:
movies.shape

(4803, 20)

In [8]:
credits_updated_df.shape

(4803, 4)

### Merging the datasets

In [9]:
# Merging the two datasets on 'id' column
# rename new data to movies_df_merge

movies_df_merge = movies.merge(credits_updated_df, on='id')

In [10]:
movies_df_merge.shape

(4803, 23)

In [11]:
movies_df_merge.head(2)

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,...,runtime,spoken_languages,status,tagline,title_x,vote_average,vote_count,title_y,cast,crew
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...",...,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800,Avatar,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...",...,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500,Pirates of the Caribbean: At World's End,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."


In [12]:
movies_df_merge.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title_x', 'vote_average',
       'vote_count', 'title_y', 'cast', 'crew'],
      dtype='object')

### Deleting Unimportant features

In [13]:
# Droping the columns which are not required for our recommendation system
# rename new data to movies_cleaned_df

movies_cleaned_df = movies_df_merge.drop(['homepage', 'title_x', 'title_y', 'production_countries'], axis=1)

<a id="1"></a>
<p style="padding:10px;
          background: linear-gradient(to right, #ab70c6, #4a006c);
          margin:0;
          color:#000000;
          font-family:newtimeroman;
          font-size:250%;
          text-align:center;
          border-radius: 15px 50px;
          overflow:hidden;
          font-weight:500">
Bulding Average Weighted Recommender System
</p>


## IMDB Weighted Rating Formula

$$W = \frac{Rv + Cm}{v + m}$$

**Where:**

- $W$ = Weighted Rating 
- $R$ = average for the movie as a number from 0 to 10 (mean) = (Rating)
- $v$ = number of votes for the movie = (votes)  
- $m$ = minimum votes required to be listed in the Top 250 (currently 3000)
- $C$ = the mean vote across the whole report (currently 6.9)

In [14]:
movies_cleaned_df.columns

Index(['budget', 'genres', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'release_date', 'revenue', 'runtime', 'spoken_languages', 'status',
       'tagline', 'vote_average', 'vote_count', 'cast', 'crew'],
      dtype='object')

In [15]:
v = movies_cleaned_df['vote_count']
R = movies_cleaned_df['vote_average']
C = movies_cleaned_df['vote_average'].mean()

In [16]:
movies_cleaned_df[['original_title', 'vote_count']]

,original_title,vote_count
0,Avatar,11800
1,Pirates of the Caribbean: At World's End,4500
2,Spectre,4466
3,The Dark Knight Rises,9106
4,John Carter,2124
...,...,...
4798,El Mariachi,238
4799,Newlyweds,5
4800,"Signed, Sealed, Delivered",6
4801,Shanghai Calling,7


In [17]:
px.box(movies_cleaned_df, x='vote_count', title='Box Plot of Vote Count')

m = 1900

In [18]:
threshold = movies_cleaned_df['vote_count'].quantile(q=0.9)
threshold

np.float64(1838.4000000000015)

In [19]:
movies_cleaned_df['vote_count'] >= threshold

0        True
1        True
2        True
3        True
4        True
        ...  
4798    False
4799    False
4800    False
4801    False
4802    False
Name: vote_count, Length: 4803, dtype: bool

In [20]:
filtered_df = movies_cleaned_df[movies_cleaned_df['vote_count'] >= threshold]
filtered_df.head(3)

,budget,genres,id,keywords,original_language,original_title,overview,popularity,production_companies,release_date,revenue,runtime,spoken_languages,status,tagline,vote_average,vote_count,cast,crew
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,7.2,11800,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",6.9,4500,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...",2015-10-26,880674609,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,6.3,4466,"[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."


In [21]:
# R, v, C, m

In [22]:
m = threshold
m

np.float64(1838.4000000000015)

In [23]:
R

0       7.2
1       6.9
2       6.3
3       7.6
4       6.1
       ... 
4798    6.6
4799    5.9
4800    7.0
4801    5.7
4802    6.3
Name: vote_average, Length: 4803, dtype: float64

In [24]:
C

np.float64(6.092171559442016)

In [25]:
v

0       11800
1        4500
2        4466
3        9106
4        2124
        ...  
4798      238
4799        5
4800        6
4801        7
4802       16
Name: vote_count, Length: 4803, dtype: int64

In [26]:
v = filtered_df['vote_count']
R = filtered_df['vote_average']
C = movies_cleaned_df['vote_average'].mean()

In [27]:
v

0       11800
1        4500
2        4466
3        9106
4        2124
        ...  
4291     2184
4300     3697
4302     2311
4337     2535
4602     2078
Name: vote_count, Length: 481, dtype: int64

In [28]:
C

np.float64(6.092171559442016)

In [29]:
R

0       7.2
1       6.9
2       6.3
3       7.6
4       6.1
       ... 
4291    7.2
4300    8.0
4302    8.1
4337    8.0
4602    8.2
Name: vote_average, Length: 481, dtype: float64

In [30]:
filtered_df['weighted_avg'] = ((R*v)+(C*m))/(v+m)
filtered_df.head(3)

C:\Users\CM\AppData\Local\Temp\ipykernel_6336\3762658069.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



,budget,genres,id,keywords,original_language,original_title,overview,popularity,production_companies,release_date,revenue,runtime,spoken_languages,status,tagline,vote_average,vote_count,cast,crew,weighted_avg
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,7.2,11800,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de...",7.050669
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",6.9,4500,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de...",6.665696
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...",2015-10-26,880674609,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,6.3,4466,"[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de...",6.239396


Top 20 most recommended movies based on average weighted :

In [31]:
df_sorted_ranking = filtered_df.sort_values('weighted_avg', ascending=False).head(20)
df_sorted_ranking.head(3)

,budget,genres,id,keywords,original_language,original_title,overview,popularity,production_companies,release_date,revenue,runtime,spoken_languages,status,tagline,vote_average,vote_count,cast,crew,weighted_avg
1881,25000000,"[{""id"": 18, ""name"": ""Drama""}, {""id"": 80, ""name...",278,"[{""id"": 378, ""name"": ""prison""}, {""id"": 417, ""n...",en,The Shawshank Redemption,Framed in the 1940s for the double murder of h...,136.747729,"[{""name"": ""Castle Rock Entertainment"", ""id"": 97}]",1994-09-23,28341469,142.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,Fear can hold you prisoner. Hope can set you f...,8.5,8205,"[{""cast_id"": 3, ""character"": ""Andy Dufresne"", ...","[{""credit_id"": ""52fe4231c3a36847f800b153"", ""de...",8.059258
662,63000000,"[{""id"": 18, ""name"": ""Drama""}]",550,"[{""id"": 825, ""name"": ""support group""}, {""id"": ...",en,Fight Club,A ticking-time-bomb insomniac and a slippery s...,146.757391,"[{""name"": ""Regency Enterprises"", ""id"": 508}, {...",1999-10-15,100853753,139.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,Mischief. Mayhem. Soap.,8.3,9413,"[{""cast_id"": 4, ""character"": ""The Narrator"", ""...","[{""credit_id"": ""55731b8192514111610027d7"", ""de...",7.939256
65,185000000,"[{""id"": 18, ""name"": ""Drama""}, {""id"": 28, ""name...",155,"[{""id"": 849, ""name"": ""dc comics""}, {""id"": 853,...",en,The Dark Knight,Batman raises the stakes in his war on crime. ...,187.322927,"[{""name"": ""DC Comics"", ""id"": 429}, {""name"": ""L...",2008-07-16,1004558444,152.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Why So Serious?,8.2,12002,"[{""cast_id"": 35, ""character"": ""Bruce Wayne"", ""...","[{""credit_id"": ""55a0eb4a925141296b0010f8"", ""de...",7.920020


In [32]:
df_sorted_ranking.columns

Index(['budget', 'genres', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'release_date', 'revenue', 'runtime', 'spoken_languages', 'status',
       'tagline', 'vote_average', 'vote_count', 'cast', 'crew',
       'weighted_avg'],
      dtype='object')

In [33]:
df_sorted_ranking[['original_title', 'vote_count', 'vote_average', 'weighted_avg', 'popularity']]

,original_title,vote_count,vote_average,weighted_avg,popularity
1881,The Shawshank Redemption,8205,8.5,8.059258,136.747729
662,Fight Club,9413,8.3,7.939256,146.757391
65,The Dark Knight,12002,8.2,7.920020,187.322927
3232,Pulp Fiction,8428,8.3,7.904645,121.463076
96,Inception,13752,8.1,7.863239,167.583710
3337,The Godfather,5893,8.4,7.851236,143.659698
95,Interstellar,10867,8.1,7.809479,724.247784
809,Forrest Gump,7927,8.2,7.803188,138.133331
329,The Lord of the Rings: The Return of the King,8064,8.1,7.727243,123.630332
1990,The Empire Strikes Back,5879,8.2,7.697884,78.517830


In [34]:
import warnings
from warnings import filterwarnings
filterwarnings('ignore')

In [35]:
df_sorted_ranking.columns

Index(['budget', 'genres', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'release_date', 'revenue', 'runtime', 'spoken_languages', 'status',
       'tagline', 'vote_average', 'vote_count', 'cast', 'crew',
       'weighted_avg'],
      dtype='object')

In [36]:
fig_weighted_avg = px.bar(
                    df_sorted_ranking,
                    x='weighted_avg',
                    y='original_title',
                    orientation='h',
                    color='weighted_avg',
                    color_continuous_scale=['#bda0ca', '#4a006c']
                    )

fig_weighted_avg.update_layout(yaxis={'categoryorder':'total ascending'})

fig_weighted_avg.show()


#### 🎬 Insight:
- The Shawshank Redemption ranks #1 with a weighted average above 8.1
- All 21 movies score between 7.6–8.1, dominated by thrillers and dramas
- The Lord of the Rings is the most represented franchise with 3 entries

<a id="1"></a>
<p style="padding:10px;
          background: linear-gradient(to right, #ab70c6, #4a006c);
          margin:0;
          color:#000000;
          font-family:newtimeroman;
          font-size:250%;
          text-align:center;
          border-radius: 15px 50px;
          overflow:hidden;
          font-weight:500">
Building Popularity-based Recommender System
</p>


In [37]:
filtered_df.columns

Index(['budget', 'genres', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'release_date', 'revenue', 'runtime', 'spoken_languages', 'status',
       'tagline', 'vote_average', 'vote_count', 'cast', 'crew',
       'weighted_avg'],
      dtype='object')

In [38]:
popularity_df = filtered_df.sort_values('popularity', ascending=False).head(20)
popularity_df.head(3)

,budget,genres,id,keywords,original_language,original_title,overview,popularity,production_companies,release_date,revenue,runtime,spoken_languages,status,tagline,vote_average,vote_count,cast,crew,weighted_avg
546,74000000,"[{""id"": 10751, ""name"": ""Family""}, {""id"": 16, ""...",211672,"[{""id"": 3487, ""name"": ""assistant""}, {""id"": 179...",en,Minions,"Minions Stuart, Kevin and Bob are recruited by...",875.581305,"[{""name"": ""Universal Pictures"", ""id"": 33}, {""n...",2015-06-17,1156730962,91.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"Before Gru, they had a history of bad bosses",6.4,4571,"[{""cast_id"": 22, ""character"": ""Scarlet Overkil...","[{""credit_id"": ""5431b2b10e0a2656e20026c7"", ""de...",6.311706
95,165000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 18, ""...",157336,"[{""id"": 83, ""name"": ""saving the world""}, {""id""...",en,Interstellar,Interstellar chronicles the adventures of a gr...,724.247784,"[{""name"": ""Paramount Pictures"", ""id"": 4}, {""na...",2014-11-05,675120017,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,Mankind was born on Earth. It was never meant ...,8.1,10867,"[{""cast_id"": 9, ""character"": ""Joseph Cooper"", ...","[{""credit_id"": ""52fe4bbf9251416c910e4801"", ""de...",7.809479
788,58000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",293660,"[{""id"": 2095, ""name"": ""anti hero""}, {""id"": 307...",en,Deadpool,Deadpool tells the origin story of former Spec...,514.569956,"[{""name"": ""Twentieth Century Fox Film Corporat...",2016-02-09,783112979,108.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,Witness the beginning of a happy ending,7.4,10995,"[{""cast_id"": 99, ""character"": ""Wade Wilson / D...","[{""credit_id"": ""56c986b2925141172f0068b6"", ""de...",7.212652


In [39]:
fig_popularity = px.bar(
                    popularity_df,
                    x='popularity',
                    y='original_title',
                    orientation='h',
                    color='popularity',
                    color_continuous_scale=['#bda0ca', '#4a006c']
                    )

fig_popularity.update_layout(yaxis={'categoryorder':'total ascending'})

fig_popularity.show()

#### 🎬 Insight:
- Minions dominates popularity with a score near 900, far ahead of Interstellar (~700)
- Most movies cluster below 300, showing a large popularity gap between blockbusters and the rest

<p style="padding:10px;
          background: linear-gradient(to right, #ab70c6, #4a006c);
          margin:0;
          color:#000000;
          font-family:newtimeroman;
          font-size:250%;
          text-align:center;
          border-radius: 15px 50px;
          overflow:hidden;
          font-weight:500">
Normalize the Data <br>
<span style="font-size:60%;">(for Hybrid Model)</span>
</p>

In [40]:
filtered_df.shape

(481, 20)

In [41]:
filtered_df.columns

Index(['budget', 'genres', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'release_date', 'revenue', 'runtime', 'spoken_languages', 'status',
       'tagline', 'vote_average', 'vote_count', 'cast', 'crew',
       'weighted_avg'],
      dtype='object')

In [42]:
from sklearn.preprocessing import MinMaxScaler

In [43]:
scaling = MinMaxScaler()

In [44]:
filtered_df[['weighted_avg', 'popularity']][0:5]

,weighted_avg,popularity
0,7.050669,150.437577
1,6.665696,139.082615
2,6.239396,107.376788
3,7.346721,112.312950
4,6.096368,43.926995


In [45]:
scaling_values = scaling.fit_transform(filtered_df[['weighted_avg', 'popularity']])

In [46]:
scaling_values[0:5]

array([[0.65263343, 0.17005206],
       [0.52004532, 0.15705597],
       [0.37322387, 0.12076774],
       [0.75459604, 0.12641732],
       [0.32396375, 0.04814762]])

In [47]:
filtered_df[['weighted_avg_scaled', 'popularity_scaled']] = scaling_values

In [48]:
filtered_df.head(3)

,budget,genres,id,keywords,original_language,original_title,overview,popularity,production_companies,release_date,...,spoken_languages,status,tagline,vote_average,vote_count,cast,crew,weighted_avg,weighted_avg_scaled,popularity_scaled
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...",2009-12-10,...,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,7.2,11800,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de...",7.050669,0.652633,0.170052
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...",2007-05-19,...,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",6.9,4500,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de...",6.665696,0.520045,0.157056
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...",2015-10-26,...,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,6.3,4466,"[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de...",6.239396,0.373224,0.120768


<p style="padding:10px;
          background: linear-gradient(to right, #ab70c6, #4a006c);
          margin:0;
          color:#000000;
          font-family:newtimeroman;
          font-size:250%;
          text-align:center;
          border-radius: 15px 50px;
          overflow:hidden;
          font-weight:500">
Build Hybrid Recommendation Feature <br>
<span style="font-size:50%;">(based on weighted average & Popularity score)</span>
</p>

In [49]:
filtered_df.columns

Index(['budget', 'genres', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'release_date', 'revenue', 'runtime', 'spoken_languages', 'status',
       'tagline', 'vote_average', 'vote_count', 'cast', 'crew', 'weighted_avg',
       'weighted_avg_scaled', 'popularity_scaled'],
      dtype='object')

In [50]:
filtered_df['score_mix'] = filtered_df['weighted_avg_scaled']*0.5 + filtered_df['popularity_scaled']*0.5

In [51]:
filtered_df.head(2)

,budget,genres,id,keywords,original_language,original_title,overview,popularity,production_companies,release_date,...,status,tagline,vote_average,vote_count,cast,crew,weighted_avg,weighted_avg_scaled,popularity_scaled,score_mix
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...",2009-12-10,...,Released,Enter the World of Pandora.,7.2,11800,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de...",7.050669,0.652633,0.170052,0.411343
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...",2007-05-19,...,Released,"At the end of the world, the adventure begins.",6.9,4500,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de...",6.665696,0.520045,0.157056,0.338551


In [52]:
recommended_movies = filtered_df.sort_values('score_mix', ascending=False)[0:20]
recommended_movies.head(3)

,budget,genres,id,keywords,original_language,original_title,overview,popularity,production_companies,release_date,...,status,tagline,vote_average,vote_count,cast,crew,weighted_avg,weighted_avg_scaled,popularity_scaled,score_mix
95,165000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 18, ""...",157336,"[{""id"": 83, ""name"": ""saving the world""}, {""id""...",en,Interstellar,Interstellar chronicles the adventures of a gr...,724.247784,"[{""name"": ""Paramount Pictures"", ""id"": 4}, {""na...",2014-11-05,...,Released,Mankind was born on Earth. It was never meant ...,8.1,10867,"[{""cast_id"": 9, ""character"": ""Joseph Cooper"", ...","[{""credit_id"": ""52fe4bbf9251416c910e4801"", ""de...",7.809479,0.913974,0.826794,0.870384
546,74000000,"[{""id"": 10751, ""name"": ""Family""}, {""id"": 16, ""...",211672,"[{""id"": 3487, ""name"": ""assistant""}, {""id"": 179...",en,Minions,"Minions Stuart, Kevin and Bob are recruited by...",875.581305,"[{""name"": ""Universal Pictures"", ""id"": 33}, {""n...",2015-06-17,...,Released,"Before Gru, they had a history of bad bosses",6.4,4571,"[{""cast_id"": 22, ""character"": ""Scarlet Overkil...","[{""credit_id"": ""5431b2b10e0a2656e20026c7"", ""de...",6.311706,0.398128,1.000000,0.699064
94,170000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 878, ""na...",118340,"[{""id"": 8828, ""name"": ""marvel comic""}, {""id"": ...",en,Guardians of the Galaxy,"Light years from Earth, 26 years after being a...",481.098624,"[{""name"": ""Marvel Studios"", ""id"": 420}, {""name...",2014-07-30,...,Released,All heroes start somewhere.,7.9,9742,"[{""cast_id"": 1, ""character"": ""Peter Quill / St...","[{""credit_id"": ""538ce329c3a3687155003358"", ""de...",7.613005,0.846307,0.548503,0.697405


In [53]:
print(recommended_movies[['score_mix']].head())

     score_mix
95    0.870384
546   0.699064
94    0.697405
788   0.647617
65    0.582157


In [54]:
fig_mix = px.bar(
            recommended_movies,
            x='score_mix',
            y='original_title',
            orientation='h',
            color='score_mix',
            color_continuous_scale=['#bda0ca', '#4a006c']
            )

fig_mix.update_layout(yaxis={'categoryorder':'total ascending'})

fig_mix.show()

#### 🎬 Insight:
- Interstellar tops the combined score_mix (~0.88), balancing both high rating and popularity
- Minions and Guardians of the Galaxy follow closely, showing that mainstream blockbusters
- score well when both popularity and rating are combined

##### This plot shows the weighted average, popularity, and mix score for each movie :

In [55]:
import plotly.express as px

fig = px.bar(
    recommended_movies,
    y='original_title',
    x='score_mix',
    orientation='h',
    color='score_mix',
    color_continuous_scale=['#bda0ca', '#4a006c'],

    hover_data={
        'weighted_avg': True,
        'popularity': True,
        'score_mix': True,
        'original_title': False
    }
)

fig.update_layout(
    yaxis={'categoryorder': 'total ascending'},
    # xaxis_tickangle=65,
    width=900,
    height=600
)

fig.show()

#### 🎬 Insight:
- Interstellar leads with the highest score_mix (~0.88), combining strong ratings and popularity
- Scores drop gradually from ~0.65 to 0.50, suggesting most recommended movies are closely matched

<a id="1"></a>
<p style="padding:10px;
          background: linear-gradient(to right, #ab70c6, #4a006c);
          margin:0;
          color:#000000;
          font-family:newtimeroman;
          font-size:250%;
          text-align:center;
          border-radius: 15px 50px;
          overflow:hidden;
          font-weight:500">
Results
</p>


#### - Interstellar tops the recommendation list with the highest score_mix (approximately 0.88). This confirms it perfectly balances a very high rating with strong popularity.
#### - Minions and Guardians of the Galaxy come in second and third, showing that light, entertaining blockbuster movies perform exceptionally well when popularity is combined with ratings.
#### - The Dark Knight and The Shawshank Redemption remain among the highest-rated films even years after their release, proving the lasting power of classic movies in the recommendation system.
#### - Movies that achieve a strong balance between high ratings and high popularity (such as Inception, Fight Club, and Mad Max: Fury Road) consistently rank at the top when using Weighted Average + Popularity together.
#### - The score_mix varies significantly across movies, starting from 0.88 for the best film and gradually dropping to around 0.50. This indicates that the recommendations are diverse and not limited to just one genre or type.
#### - High-rated and highly popular films like Star Wars and The Lord of the Rings receive high priority in recommendations, making the system much smarter than relying on rating alone.
#### - The system proves effective at recommending a wide variety of movies (Drama, Action, Animation, Sci-Fi) as long as they are strong in both rating and popularity.